### PR TITLE
[WIP] initial implementation to support audio processing as arrays

### PR DESCRIPTION
**Pull request by g8a9**
This PR has multiple goals. 
The main one is to **support using numpy arrays for audio rather than local paths only.** I believe this could be a strong limitation since many audio datasets today come from the HF Hub already decoded into np.ndarray.

However, while inspecting the code, I've noticed there are still **several dependencies from local files that we don't want to have.** For example, the code expects:
- local transcripts for most of the explainers (I think we should have a centralized place where we compute them)
- local files for adding noise (white and pink)

Another important goal is to **standardize transcript generation,** which is currently made only by the LOO explainers.

### CODE REVIEW

**Comment by g8a9** [comment]
@emanuele-moscato I've left around several todo markers for points to discuss. You can find them with `#TODO GA`

**Comment by emanuele-moscato** [comment]
Look at the explainers under `ferret/explainers/explanation_speech/`:

- [x] Whenever the path to an audio file is required (typically in the `compute_explanation` methods), pass an object of type `FerretAudio` (newly defined in the PR).
- [x] Make `FerretAudio` able to transcribe to text (dedicated method called from within the explainers when/if needed). --> NOT VALID ANYMORE
- [x] Check the SpeechXAI examples notebook.

File: ferret/explainers/explanation_speech/gradient_speech_explainer.py [hunk 1]
```diff
@@ -181,7 +177,7 @@ def compute_explanation(
             scores=scores,
             explainer=self.NAME + "-" + aggregation,
             target=targets if n_labels > 1 else targets,
-            audio_path=audio_path,
+            audio_path=audio,
```
Comment (emanuele-moscato) [inline ferret/explainers/explanation_speech/gradient_speech_explainer.py]: Let's change the name of this optional argument in to `audio`, for consistency. This needs to be done in the `ExplanationSpeech` class within `explanation_speech.py`.

File: ferret/explainers/explanation_speech/lime_speech_explainer.py [hunk 2]
```diff
@@ -143,7 +140,7 @@ def compute_explanation(
             scores=scores,
             explainer=self.NAME + "+" + removal_type,
             target=targets if n_labels > 1 else targets,
-            audio_path=audio_path,
+            audio_path=audio,
```
Comment (emanuele-moscato) [inline ferret/explainers/explanation_speech/lime_speech_explainer.py]: As in the other explainers: change keyword argument's name to `audio` in the `ExplanationSpeech` class.

File: ferret/explainers/explanation_speech/loo_speech_explainer.py [hunk 3]
```diff
@@ -1,11 +1,15 @@
...
 from pydub import AudioSegment
 from IPython.display import display
 from .explanation_speech import ExplanationSpeech
-from .utils_removal import transcribe_audio, remove_word
-from ...speechxai_utils import pydub_to_np, print_log
+from .utils_removal import remove_word, remove_word_np
+from ...speechxai_utils import pydub_to_np, FerretAudio
+from logging import getLogger
+
+logger = getLogger(__name__)
```
Comment (emanuele-moscato) [inline ferret/explainers/explanation_speech/loo_speech_explainer.py:12]: We are starting to use loggers in different places: let's use a singleton pattern so we avoid having multiple ones instantiated at the same time.

File: ferret/explainers/explanation_speech/loo_speech_explainer.py [hunk 4]
```diff
@@ -125,7 +127,7 @@ def compute_explanation(
             scores=scores,
             explainer=self.NAME + "+" + removal_type,
             target=targets if n_labels > 1 else [targets],
-            audio_path=audio_path,
+            audio=audio,  # TODO GA: I don't know if this is something we want to keep
```
Comment (emanuele-moscato) [inline ferret/explainers/explanation_speech/loo_speech_explainer.py:123]: As above, but the keyword argument's name here has been updated already, but now I'm confused: the class hasn't been modified!

File: ferret/speechxai_utils.py [hunk 5]
```diff
@@ -7,6 +7,81 @@
...
+    
+    @property
+    def normalized_array(self) -> np.ndarray:
+        if not self.is_normalized:
+            return self.array / 32768.0
+        else:
+            return self.array
+
+    @property
+    def transcribe(self):
```
Comment (emanuele-moscato) [inline ferret/speechxai_utils.py]: I would have called this `transcription` as it's used as an attribute - but I'm aware I'm being pedantic here ahah!

File: ferret/speechxai_utils.py [hunk 6]
```diff
@@ -7,6 +7,81 @@
 import torch
 from datasets import Dataset
 from transformers import Wav2Vec2ForSequenceClassification, Wav2Vec2FeatureExtractor
+import librosa
+from typing import Union
+from explainers.explanation_speech.utils_removal import transcribe_audio
```
Comment (emanuele-moscato) [inline ferret/speechxai_utils.py]: Watch out: if I try to import stuff from the current module (e.g. the `FerretAudio` class), I get an error because the `explainers` is not found: we need a relative import here (`from .explainers....`). But now `utils_removal.py` tries to import from `speechxai_utils.py`, generating a circular import.

**Comment by emanuele-moscato** [comment]
To do:
- [x] In the definition of the `ExplanationSpeech` class, change the kwarg name from `audio_path` to `audio` and fix code accordingly everywhere the class is implemented (make sure an object of type `FerretAudio` is passed).
- [x] Address the comments in the last code review.
- [x] Complete the update of the speech explainers so they accept an object of type `FerretAudio` as their input.
- [x] See previous to-do list.

**Note:** the audio transcription functions `transcribe_audio` and `transcribe_audio_with_model` were moved from the `ferret/explainers/explanation_speech/utils_removal.py` module to `ferret/speechxai_utils.py` to avoid a circular import. The rest of the code has already been updated, but have one last check please!

File: ferret/explainers/explanation_speech/equal_width/gradient_equal_width_explainer.py [hunk 7]
```diff
@@ -117,7 +113,7 @@ def compute_explanation(
             importances = []
             a, b = 0, 0  # 50, 20
 
-            duration_s = len(audio_as) / 1000
+            duration_s = len(audio_array) / audio.sample_rate # finds the duration from the array
```
Comment (emanuele-moscato) [inline ferret/explainers/explanation_speech/equal_width/gradient_equal_width_explainer.py]: Let's double-check this: the original factor `1000` was converting fro milliseconds to seconds (so `audio_as` was assumed to be an array with lengths equal the duration in milliseconds) while dividing by the sample rate requires that `audio_array` is now as long as the number of samples for a frequency equals to the sample rate (in hz).

File: ferret/explainers/explanation_speech/equal_width/lime_equal_width_explainer.py [hunk 8]
```diff
@@ -59,13 +56,13 @@ def compute_explanation(
...
 
-        audio_np = audio.reshape(1, -1)
+        audio_np = audio_array.reshape(1, -1)
 
         # Get the start and end indexes of the segments. These will be used to split the audio and derive LIME interpretable features
         sampling_rate = self.model_helper.feature_extractor.sampling_rate
         splits = []
 
-        duration_s = len(audio_as) / 1000
+        duration_s = len(audio_array) / audio.sample_rate # finds the duration from the array
```
Comment (emanuele-moscato) [inline ferret/explainers/explanation_speech/equal_width/lime_equal_width_explainer.py]: Same double-check as in `gradient_equal_width_explainer.py`.

File: ferret/explainers/explanation_speech/equal_width/loo_equal_width_explainer.py [hunk 9]
```diff
@@ -65,19 +63,17 @@ def compute_explanation(
...
-        ## Load audio as pydub.AudioSegment
-        audio = AudioSegment.from_wav(audio_path)
-        audio_np = pydub_to_np(audio)[0]
+        audio_array = audio.array
 
         ## Remove word
         audio_remove_segments = []
 
-        duration_s = len(audio) / 1000
+        duration_s = len(audio_array) / audio.sample_rate # finds the duration from the array
```
Comment (emanuele-moscato) [inline ferret/explainers/explanation_speech/equal_width/loo_equal_width_explainer.py]: Same double-check as in `gradient_equal_width_explainer.py`.

**Comment by emanuele-moscato** [comment]
Issues to solve:
- [x] When loading an audio file into a `FerretAudio` object and then extracting the `AudioSegment` object (`to_pydub` method), the audio gets distorted. Check `FerretAudio`'s `to_pydub` method.

**Action:** modified `FerretAudio`'s `to_pydub` method so as to create a pudyb `AudioSegment` object from an array that is always unnormalized and of dtype `int16`.

- [x] Move audio resampling (if the native sample rate is different from 16 KHz) to within the transcription method.
- [x] Move normalization of the array corresponding to the audio (if needed) to within the transcription method as well (so there are no ambiguities on whether the array provided by the user is normalized or not: it is what it is from the start).
- [x] Check that the new way of obtaining arrays from audio (`librosa.load`) returns the same as (`AudioSegment.load_wav` --> `pydub_to_np`).

**Result:** numerically it is, they both return an array of dtype float32 normalized by a factor 2 ** 15, but the shape is different as (for mono audio) `librosa` returns a flattened array while `pydub_to_np` returns an array of shape `(n_samples, 1)`.

- [x] Raise error if the numpy array or audio file passed to `FerretAudio` has more than one channel (**we only support mono audio!**).

**Action:** this is inferred looking at the shape of the array. 1-dimensional array --> good, 2-dimensional array --> good only if the trailing dimension is 1 (shape `(n_samples, 1)`), array of dimension > 2 --> shape not understood, raise exception.

- [x] Remove the `remove_word_np` function if it ends up not being used.

File: ferret/explainers/explanation_speech/equal_width/loo_equal_width_explainer.py (hunk 9 above)
Comment (g8a9) [reply ferret/explainers/explanation_speech/equal_width/loo_equal_width_explainer.py]: the denominator `1000` assumed to work with milliseconds. Is this change still valid? @emanuele-moscato

**Comment by g8a9** [comment]
Regarding the normalization: we should not touch the input array unless 1) it's not normalized and 2) we are transcribing it with whisperX -- that's the only case where we really need normalization + 16kHz sampling rate.

File: ferret/explainers/explanation_speech/equal_width/gradient_equal_width_explainer.py (hunk 7 above)
Comment (g8a9) [reply ferret/explainers/explanation_speech/equal_width/gradient_equal_width_explainer.py]: @emanuele-moscato we should check this one, too
